go-selinux/selinux_linux.go

```diff
@@ -1095,3 +1104,115 @@ func dupSecOpt(src string) ([]string, error) {
 func disableSecOpt() []string {
 	return []string{"disable"}
 }
+
+func getDefaultContextFromReaders(c defaultSECtx) (string, error) {
+	if c.verifier == nil {
+		return "", fmt.Errorf("verifier function must not be nil")
+	}
+
+	context, err := newContext(c.scon)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to create label for %s", c.scon)
+	}
+
+	findUserInContext := func(r io.Reader) string {
+		fromRole := context["role"]
+		fromType := context["type"]
+		bufin := bufio.NewReader(r)
+
+		for done := false; !done; {
+			var (
+				line string
+				err  error
+			)
+
+			if line, err = bufin.ReadString('\n'); err != nil {
+				if err == io.EOF {
+					done = true
+				} else {
+					break
+				}
+			}
+			line = strings.TrimSpace(line)
+			if len(line) == 0 {
+				// Skip blank lines
+				continue
+			}
+			if line[0] == ';' || line[0] == '#' {
+				// Skip comments
+				continue
+			}
+
+			conns := strings.Fields(line)
+
+			lineConn := strings.TrimSpace(conns[0])
+			lineArr := strings.SplitN(lineConn, ":", 3)
+
+			// malformed context in the context file
+			if len(lineArr) != 3 {
+				continue
+			}
+
+			if lineArr[0] != fromRole || lineArr[1] != fromType {
+				continue
+			}
+
+			for _, cc := range conns[1:] {
+				userConn := strings.TrimSpace(cc)
+				userArr := strings.SplitN(userConn, ":", 3)
+
+				context["user"] = c.user
+				context["role"] = userArr[0]
+				context["type"] = userArr[1]
+				context["level"] = c.level
+
+				outConn := context.get()
+				if err := c.verifier(outConn); err != nil {
+					continue
+				}
+
+				return outConn
+			}
+		}
+
+		return ""
+	}
+
+	if conn := findUserInContext(c.userRdr); conn != "" {
+		return conn, nil
+	}
+
+	if conn := findUserInContext(c.defaultRdr); conn != "" {
+		return conn, nil
+	}
+
+	return "", fmt.Errorf("could not find context match for %q", c.scon)
+}
+
+func getDefaultContextWithLevel(user, level, scon string) (string, error) {
+	userPath := filepath.Join(getSELinuxPolicyRoot(), selinuxUsersDir, user)
+	defaultPath := filepath.Join(getSELinuxPolicyRoot(), defaultContexts)
```

Comment:
Funny that every line here causes re-reading of `/etc/selinux/config`. At the very least you can reuse the `getSELinuxPolicyRoot` result here.

Feel free to optimize that in the existing code as well (separate commit).